### PR TITLE
✨(backend) add ServiceProvider

### PR DESCRIPTION
**Pull request by qbey**
## Purpose

We want to allow Service Providers to fetch our data but we also need a better management of the data available for each one of them.

## Proposal

Not included in this PR: allow Service Provider management from the front.

⚠️ This disables the resource server globaly, to only allow it on a per-view basis because we don't want to allow any service provider to be able to query everything without filtering.

- [x] add the `ServiceProvider` model
- [x] allow Service Provider management from Django admin
- [x] restrict endpoint availability as a resource server 
- [x] update changelog

### CODE REVIEW

File: src/backend/core/tests/service_providers/test_core_api_service_providers_retrieve.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    response = client.get(f"/api/v1.0/service-providers/{service_provider.pk}/")
+
+    assert response.status_code == HTTP_404_NOT_FOUND
+    assert response.json() == {"detail": "No ServiceProvider matches the given query."}
+
+
+def test_api_service_providers_retrieve_authenticated_on_teams(client):
+    """
+    Authenticated users should not be able to retrieve service providers
+    of because of their teams (might change later if needed).
```
Comment (sampaccoud) [inline src/backend/core/tests/service_providers/test_core_api_service_providers_retrieve.py:68]:
```suggestion
    just because it is related to one of their teams if it is not related to their organization (might change later if needed).
```

File: src/backend/core/tests/teams/test_core_api_teams_update.py [hunk 2]
```diff
@@ -188,3 +190,95 @@ def test_api_teams_update_administrator_or_owner_of_another():
     team.refresh_from_db()
     team_values = serializers.TeamSerializer(instance=team).data
     assert team_values == old_team_values
+
+
+@responses.activate
+def test_api_teams_update_authenticated_owner_resource_server(
```
Comment (sampaccoud) [inline src/backend/core/tests/teams/test_core_api_teams_update.py:196]: I was expecting tests about adding/removing service providers from a team for authenticated admin/owners of the team?

File: src/backend/core/tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            ],
+        )
+
+    yield {
+        "encrypt_jwt": encrypt_jwt,
+    }
+
+
+@pytest.fixture
+def authenticated_user_with_service_provider(client):
```
Comment (sampaccoud) [inline src/backend/core/tests/conftest.py:114]:
```suggestion
def user_authenticated_via_service_provider(client):
```

File: src/backend/core/tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
+"""Defines fixtures for the tests."""
```
Comment (sampaccoud) [inline src/backend/core/tests/conftest.py:1]: hope we don't abuse this pattern. I like tests to be self understandable without having to jump from file to file to try to understand each fixture.
+ the necessity to define edge cases tends to create complicated code when you try to mutualize test code.

**Comment by qbey** [comment]
@sampaccoud sorry I was not able to keep fixup commits as the refacto was too big to move resource server APIs to a dedicated module.
I think I fixed all your comments (except for 404 vs 403, but we agreed IRL on this).
The URL split for resource server makes the tests way cleaner :+1: and allows to have more specific endpoints, this is a good improvement :)

**Review by sampaccoud — APPROVED** [review]

File: src/backend/core/resource_server_api/serializers.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        """
+        Create a new team with organization enforcement.
+
+        In this context, called as a resource server,
+        the team service provider is enforced.
+
+        When the service provider audience is unknown it is created on the fly.
+        """
+        sp_audience = self.context["from_service_provider_audience"]
+        service_provider, _created = models.ServiceProvider.objects.get_or_create(
```
[the remark this answers is not in the document]
Comment (qbey) [reply src/backend/core/resource_server_api/serializers.py:35]: @sampaccoud Good point, it's just I didn't want to "create" object when a user make a "GET" request, that is why I restricted to the "team creation" endpoint.
You think we should (get or) create the ServiceProvider for every authentication? If you do think so, I will create an issue to implement it after this PR.

File: src/backend/core/resource_server_api/serializers.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
+"""Client serializers for the People core app resource server API."""
+
+from rest_framework import serializers
```
[the remark this answers is not in the document]
Comment (qbey) [reply src/backend/core/resource_server_api/serializers.py:3]: I was wondering what is the "cleanest" way for this. Maybe it's only the naming that refrains me to do so, how should be named the "classic" API module? I don't want to name it `api` because it would result in `core.api.api` ^^' 
Actually, I'm wondering if the whole "resource_server" part should not be moved outside the "core" module?

If you have something in mind, I would happily discuss it with you :)

File: src/backend/core/resource_server_api/serializers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (qbey) [reply src/backend/core/resource_server_api/serializers.py:35]: Discussed IRL, we keep it as it is while there is no need to duplicate the code and will re-think of it when future cases arise.

File: src/backend/core/resource_server_api/viewsets.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    """
+
+    permission_classes = [permissions.AccessPermission]
+    serializer_class = serializers.TeamSerializer
+    filter_backends = [filters.OrderingFilter]
+    ordering_fields = ["created_at"]
+    ordering = ["-created_at"]
+    queryset = models.Team.objects.all()
+    pagination_class = None
```
[the remark this answers is not in the document]
Comment (qbey) [reply src/backend/core/resource_server_api/viewsets.py:54]: Ok for pagination.

For SCIM, I will integrate https://github.com/15five/django-scim2/ later, with dedicated endpoints, because the resource server might also be used to create email address (not implemented yet) which is not an SCIM feature.
